### PR TITLE
add woke ignore traversal

### PR DESCRIPTION
**Pull request by armanrahman22**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our [guidelines](https://github.com/get-woke/woke/blob/main/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

**What is the current behavior?** (You can also link to an open issue here)
Woke currently only looks at .gitignore and .wokeignore files in the current directory.

**What is the new behavior (if this is a feature change)?**
Woke will now traverse a repo and build a priority ranked list of ignore rules in exactly the same way that git does for .gitignore files at different directory levels. The tool will look for the root git directory (where the .git folder is) and use that for the root. If no root git directory is found then woke will just run with the current directory as root (still traversing children folders).

**Does this PR introduce a breaking change?** (What changes might users need to make due to this PR?)
No

**Other information**:
Addresses #98

### CODE REVIEW

File: pkg/ignore/ignore.go
Comment (caitlinelfring) [inline pkg/ignore/ignore.go:57]: Is there ever a chance that `repo` and `err` are both `nil`? It might be worth guarding against `filesystem` never being set

**Review by caitlinelfring — COMMENTED** [review]
Excited about this contribution! I have a few concerns that I'd love to hear your thoughts on.

* Inclusion of two new dependencies seems to have increased the final binary size by 6MB (from 8 to 14)
* Duplicating of code from go-git repo -- are there any licensing concerns? Should there be a contribution to their repo to make ignore files configurable instead of copying and modifying the code in this repo?
* Performance impact - can you provide regression benchmarking for these changes on a large repo (ie large number of files and directory traversals, not large filesizes)? I'm mostly interested in memory utilization and time.
* Code coverage. There are a lot of uncovered lines added that should be tested for. You should be able to see it in codecov or the vscode golang plugin 
* I haven't looked at the details, but windows tests appear to be failing

File: go.mod
Comment (caitlinelfring) [inline go.mod:45]: These upgrades seem out of scope of this PR. If they are, can you revert them? dependabot should handle all these

File: go.mod
Comment (caitlinelfring) [inline go.mod:49]: Why use `replace` and not just import from the fork directly?

**Review by caitlinelfring — COMMENTED** [review]
This is looking much better, I just have a few comments 

(from my [original comment](https://github.com/get-woke/woke/pull/117#pullrequestreview-719687296))
> Performance impact - can you provide regression benchmarking for these changes on a large repo (ie large number of files and directory traversals, not large filesizes)? I'm mostly interested in memory utilization and time. 

Alternatively, if you could make this opt-in instead of default, this would reduce the risk of a negative impact on repos that have no use case for ignore traversal. Maybe a cli argument like `--enable-ignore-traversal`?

I also see code coverage is reduced, specifically `pkg/ignore/ignore.go` and I'd love for tests to be added to bring that back up before merging.

Lastly, this is a pretty nice feature that I'd love to have documented, a quick summary of this feature would be awesome in https://github.com/get-woke/woke/blob/main/docs/ignore.md

**Comment by KSLHacks** [comment]
Joining/subscribing to thread to stay updated :)

File: go.mod
Comment (armanrahman22) [reply go.mod:45]: fixed!

File: go.mod
Comment (armanrahman22) [reply go.mod:49]: I chose `replace` because if go-git makes changes to allow for more flexibility with ignore file types we can just switch back to using the upstream repo instead of the fork and we only have to change this one line vs the imports on each file that uses go-git. But I can also change to using direct imports if that works better!

**Comment by armanrahman22** [comment]
@caitlinelfring, @KSLHacks will be taking over any updates to this PR till I get back from vacation. I've added a section to the readme explaining the new functionality, and from the [Regression Benchmarking CI Run](https://github.com/get-woke/woke/runs/4449976195?check_suite_focus=true) it looks like it's not adding any time complexity. Added a [new benchmark](https://github.com/get-woke/woke/pull/117/commits/2653bca3d9fc21efdf88be488e19545e690341c4) to `ignore_test.go` which simulates a large repo (10,000 files and 1,000 folders) and it seems like it runs in a reasonable time. @KSLHacks will add more about the memory usage. And @KSLHacks is also looking into improving code coverage, but I think the only uncovered lines are error-catching from OS operations (not sure how to reach that in a test). Let us know if you have any more questions/concerns (and sorry this PR is taking so long haha)!

**Comment by jeremydelacruz** [comment]
Hi @caitlinelfring, I chatted with @KSLHacks and took a little bit of a look at this as well. With the benchmarks, I'm wondering what we'd like to see in terms of memory utilization metrics. My understanding is that the current benchmarks via `cob` are measuring and comparing memory allocation in `AllocedBytesPerOp`. Is this insufficient for how we'd like to measure it? If so, would love to hear your thoughts on what we'd like to see instead. Here's an example screenshot from a CI run to show what I'm referring to:
image

As for code coverage, I agree on @armanrahman22's point about the uncovered lines being mainly around OS operation error-catching. I looked into a couple potential approaches to mock these OS operations to cover error catching and increase coverage in `ignore.go`. One such approach is creating some sort of `IgnoreFactory` that can store these dependency function references as properties/fields so that they could be more easily mocked/substituted during unit testing. The `IgnoreFactory` would also be used to call the `GetRootGitDir` and `NewIgnore` methods off of. Interested in what your thoughts are on this!

**Comment by caitlinelfring** [comment]
Sincere apologies for taking so long to sync back on this. I'm not too worries about the benchmarking, if there's a real negative impact from users, we can revert this change and re-evaluate. And agreed on your points of code coverage. I don't think it's worth increasing the complexity of the tests just to get better coverage numbers in this case.

I'm good to merge this once merge conflicts have been resolved. 

Again, apologies for the delay in response. Happy to get this published soon! Thanks for your patience and all your contributions!

**Review by caitlinelfring — APPROVED** [review]

**Comment by jeremydelacruz** [comment]
No worries @caitlinelfring and thank you for your getting back to us! That all sounds good to me. I've just pushed up a commit to resolve that last merge conflict. 

Please let me know if anything else is needed to get this merged in!

**Comment by caitlinelfring** [comment]
This is now available in [`v0.18.0`](https://github.com/get-woke/woke/releases/tag/v0.18.0)

**Pull request merged**
